### ISSUE TITLE
DrawIndirectCount

### ISSUE THREAD

**Issue by frustum** [gpuweb/gpuweb#1354]
Hello,

It would be awesome if the drawIndirectCount functionality will be available in WebGPU.
DrawIndirect functionality is already available, and it's excellent, but DrawIndirectCount allows to render more advanced things.

DrawIndirectCount is available under almost all platforms now: Vulkan, OpenGL, Direct3D12, Direct3D11 with AMD AGS.
Metal API has an Argument buffer extension, which should be able to emulate such functionality.

Thank you!

`void vkCmdDrawIndirectCount(
VkBuffer buffer, VkDeviceSize offset,
 VkBuffer countBuffer, VkDeviceSize countBufferOffset,
 uint32_t maxDrawCount, uint32_t stride);`

`void vkCmdDrawIndexedIndirectCount(
VkBuffer buffer, VkDeviceSize offset,
 VkBuffer countBuffer, VkDeviceSize countBufferOffset,
 uint32_t maxDrawCount, uint32_t stride);`

**Comment by kainino0x** [comment]
Of note, our current drawIndirect/drawIndexedIndirect can only do one draw call - we don't even have a drawCount parameter like Vulkan's vkCmdDrawIndirect. This is because of the multiDrawIndirect feature which is available on [many but not all Android devices](http://vulkan.gpuinfo.org/listfeaturescore10.php?platform=android).

> VUID-vkCmdDrawIndirect-drawCount-02718
> If the multi-draw indirect feature is not enabled, drawCount must be 0 or 1

Also note the drawIndirectCount feature is required in Vulkan, and is available on [no Android devices](http://vulkan.gpuinfo.org/listdevicescoverage.php?core=1.2&feature=drawIndirectCount&platform=android).

> drawIndirectCount indicates whether the implementation supports the vkCmdDrawIndirectCount and vkCmdDrawIndexedIndirectCount functions. If this feature is not enabled, these functions must not be used.

I think we would be interested in adding optional support for both multiDrawIndirect and drawIndirectCount, but since they're optional this will be post-1.0.

**Comment by frustum** [comment]
Yep, it would be great to have that extension in the future.
multiDrawIndirectCount() is available on Android Qualcomm devices, and we are using it now.
`Device Name: Adreno (TM) 650 V@444.0 (GIT@380077c, Ifdda647016, 1598982727) (Date:09/01/20)`
So it is a simple extension in the case of the desktop browser running over Vulkan/OpenGL/Direct3D12.
Thank you

**Comment by kvark** [comment]
True! I think we can even poly-fill this everywhere. We could schedule a compute operation to happen before the render pass that would fill out a temporary GPU indirect buffer. I.e. we'd copy over the indirect data for `maxDrawCount` entries into a temporary buffer, and then zero out the counts in entries that have an index equal or above the actual count value in the buffer. Then we'd just turn a `draw_indirect_count` into a for loop of `draw_indirect` calls on platforms that need that.

Anyway, I don't think it's worth pursuing at this point, we better focus on MVP essentials instead.

**Comment by frustum** [comment]
I agree that it looks like not very important extensions. But with can make super awesome WebGPU showcases with it. CPU emulation will not work when the number of draw calls is getting more than several thousand. The driver will do the job better. Or we will wait for better times :)

**Comment by cdiggins** [comment]
> Also note the drawIndirectCount feature is required in Vulkan, and is available on [no Android devices]
(http://vulkan.gpuinfo.org/listdevicescoverage.php?core=1.2&feature=drawIndirectCount&platform=android).

Just a small note, that I actually found it on 11% of Android devices according to this search: 
http://vulkan.gpuinfo.org/listdevicescoverage.php?extension=VK_KHR_draw_indirect_count&platform=android

**Comment by kainino0x** [comment]
Ah, didn't realize it was previously an extension (still getting used to the new refactored vulkan.gpuinfo.org that splits by vulkan version).

> > Also note the drawIndirectCount feature is required in Vulkan

**Comment by alecazam** [comment]
So iOS on 5S+ has a full implementation of indirect draw, but the 63% Android score means it's limited to 1 draw call in WebGPU 1.0?   This is a call that one would only use when it's supported, and fallback to explicit draws if not.  Only having 1 draw is a reason not to use this at all.

**Comment by frustum** [comment]
Hello,

I'm asking about that particular extension because compute shaders are useless without the ability to draw the result. Or everything will be limited by a single draw call which is only suitable for simple things like terrain or grass. This is okay for CPU-driven technologies. But compute shaders allows moving everything on compute level. And boosting performance and power efficiency as a result.

This is a link to our benchmark: [https://gravitymark.com/](https://gravitymark.com/)

And I think that it must work on WebGPU. Otherwise, I don't know why the WebGPU project was initiated. We don't need much from the functionality level. The engine is perfectly working on OpenGLES 3.1 feature set level. And WebGPU should be at least equal to OpenGLES. Moreover, OpenGLES is an outdated API for everybody.

We have full WebGPU and WGSL support. And there is a minimal list of features which is stopping us from running our engine on WebGPU:

* storage buffer access from vertex and fragment shaders
* multiple viewports support (supported by all desktop API)
* multi draw indirect count (supported by all desktop API except Direct3D11 running on Nvidia/Intel, and Metal)

It is possible to workaround multi_draw_indirect_count on platforms where it's not supported by additional synchronization and loop of draw indirect command inside. And sometimes, it's working even faster than the driver.

Guys, you already did tremendous work with WebGPU. Please don't stop with it :)

**Comment by kainino0x** [comment]
AFAIK, in order to achieve compute-driven multidraw right now, you need to issue one indirect draw call for each possible draw call you want to do, then use a compute shader to generate up-to that many draw calls, zeroing out the ones you don't need. This isn't amazing, but it still might be able to achieve good performance, at least if you have a reasonable bound on the maximum number of draw calls. Unfortunately I don't know the characteristics of this technique on different architectures.

Most likely we'll later have optional features for all of those things you listed, but unlikely before MVP so we can focus on the core.

**Comment by mrshannon** [comment]
I have to 2nd the need for `multiDrawIndirect`.  In fact it should be an MVP feature.  Without it the amount of draw calls, and thus API/driver overhead, is too great for complex scenes.  And complex scenes are the only reason for many to use WebGPU over WebGL.

**Comment by frustum** [comment]
It's doable to emulate MDI with Metal. It's working much faster with Apple GPU. But all AMD is losing much performance with it. Because the driver will do a redundant job of converting draw calls back to MDI format: https://tellusim.com/metal-mdi/

**Comment by kainino0x** [comment]
If we do emulation like that we could enable it on Apple GPUs but not on AMD GPUs if it doesn't perform well! 👀

**Comment by jeremyong** [comment]
> Also note the drawIndirectCount feature is optional in Vulkan, and is available on no Android devices.

@kainino0x The GPU Info DB shows percentages based on reported device support, but this percentage isn't usable as a proxy for real world device market share. Was there any analysis done on an estimate of actual devices in the wild that don't support `multiDrawIndirect`?

**Comment by logankaser** [comment]
I think there is a potential for a performance regression compared to WebGL for renderers using https://developer.mozilla.org/en-US/docs/Web/API/WEBGL_multi_draw if this is not included.
It's only available for chrome, but in my experience it makes a large difference, perhaps because chrome's
draw call overhead for WebGL is high, and It's fairly easy even for a user of the API to implement a fallback when it's not available, which is what I did for Firefox / Safari.
For my use case, I was using it only to reduce draw call overhead.

**Comment by mmgeorge** [comment]
Just tried to use drawIndirect, if I understand correctly from @kainino0x's comment, we can currently only use an indirect buffer of u32x4 for a single draw? Would also very much like to see support for an optional drawCount parameter as that would make this much more useful than it is currently.

**Comment by JesseRMeyer** [comment]
Recording integers to an array is much faster than making a draw call.  Allowing web workers (or even later -- compute) to fill in these parameters to a shared array opens up many optimization opportunities and possibly power efficiency gains.  I strongly favor seeing a count parameter added to drawIndirect, even if for now the count value must be supplied at call time and not found in some other buffer (say, GPU local).  Otherwise, as is, this functionality is no better than supplying the draw calls directly, and possibly worse.

**Comment by lcodes** [comment]
Is there any news on this feature?

Even behind a feature flag at device creation time, it would be valuable to support this.

For many domains, ie large game worlds or CAD applications, it makes no sense to play them on mobile or older hardware anyways; the target audience already has support for it, but right now it's hidden away from wgpu while available everywhere else.

**Comment by lcodes** [comment]
Same for dispatchWorkgroupsIndirect()

The goal of a count is not to save on doing a loop on the CPU side, but to build dispatch/draw commands on the GPU side where the number of executions isn't known ahead of time. In this case, reading back the results to the CPU in order to "emulate" the drawCount or dispatchCount adds frames of delay.

This feature is the cornerstone of all modern graphics engines. In my own case it's used to generate anywhere between 100s to 1000s of compute dispatches and draw calls per frame.

**Comment by chingham** [comment]
Still not implemented ? Even as an extension, it really is a must have and is the main reason we would use WebGPU instead of WebGL

**Comment by ringoz** [comment]
we need at least MultiDrawIndirect for GPU-driven rendering.
it is possible to emulate bindless resources, but impossible to efficiently emulate MDI.

**Comment by saydric** [comment]
I totally agree, MultiDrawIndirect + Bindless textures are a must have if we want WebGPU to perform as well as a desktop app on scene with a large amount of meshes (e.g. CAD data)

**Comment by cdiggins** [comment]
We need this feature in the AEC (Architecture Engineering and Construction) industry to enable rendering of medium to large-size architectural models (aka BIM for Building Information Models) on browsers. Right now, only small models can be rendered with reasonable FPS.

**Comment by logankaser** [comment]
Bindless textures can be emulated with array textures, but MD / MDI can't

**Comment by kainino0x** [comment]
We clearly need to reevaluate the priority here so moving to M1 for us to at least re-triage.

**Comment by kainino0x** [comment]
Oops, just realized I had a duplicated bug in #4349. Let me move this here:

---

_Originally posted by @mrshannon in https://github.com/gpuweb/gpuweb/issues/1949_

> Indirect drawing with multiple indirect drawing commands is a common technique for drawing complex scenes that would otherwise be infeasible due to either an excessive number of CPU issued draw calls or scene complexity that cannot be built by the CPU alone.  This is done by:
> 
> * Executing multiple draws with a single API call.
> * Allowing the GPU to generate both geometry and the draws necessary to render it.
> * Culling out unnecessary draw calls on the GPU in more complex scenes than CPU culling could achieve.
> 
> This PR addresses adding a `multi-draw-indirect` feature.  In particular it addresses adding:
> 
> *  `multiDrawIndirect` and `multiDrawIndexedIndirect` methods on `GPURenderEncoderBase`.
>     * Allows submitting multiple draws with a single API call (multi-draw).
>     * Allows the GPU to determine the number of draw calls (draw count).
>     * Use cases:
>         * GPU derived scene data
>         * GPU based culling
>         * GPU based LOD
>         * Efficient execution of complex scenes with a large number of draws
> * Non-zero `firstInstance` for `drawIndirect`, `drawIndexedIndirect`, `multiDrawIndirect`, and `multiDrawIndexedIndirect`.
>     * This is the only available per draw input, without rebinds, that is readable in the shader.
>     * Use cases:
>         * Select instance stride vertex data
>         * Index into per object or per draw data in storage buffers
>         * Multi material, single API call, rendering
> 
> 
> # Compatibility
> 
> The required backend features to implement `multi-draw-indirect` are available on:
> 
> * Newer Apple devices (~2016+)
> * All DX12 devices
> * All Vulkan capable desktops (with up to date drivers)
> * 30% of Android devices
> 
> See the sections below for details.
> 
> 
> ## Vulkan
> 
> ### Multi-Draw
> 
> Requires the 0 or 1 restriction on the `drawCount` argument of `vkCmdDrawIndirect` and `vkCmdDrawIndexIndirect` to be relaxed to any non-negative integer.  This requires the `multiDrawIndirect` feature which is supported on:
> 
> * 99% of desktop GPUs
> * 63% of Android devices
> 
> _NOTE: The `stride` argument will always be set for tight packing, in order to maintain compatibility with DX12._
> 
> 
> ### Draw Count
> 
> Requires the `vkCmdDrawIndirectCount` and `vkCmdDrawIndexedIndirectCount` functions which are provided by either the `drawIndirectCount` feature of Vulkan 1.2 or one of the following extensions:
> 
> * `VK_AMD_draw_indirect_count`
> * `VK_KHR_draw_indirect_count`
> 
> Because `drawIndirectCount` was introduced in driver updates the statistics at [https://vulkan.gpuinfo.org](https://vulkan.gpuinfo.org) cannot be relied upon.  The following is based on the oldest card that supports `drawIndirectCount` from each manufacturer, if newer cards dropped support for `drawIndirectCount` that is not captured here.
> 
> * Intel integrated cards (that support Vulkan) support `drawIndirectCount`.
> * NVIDIA cards going back to Kepler support `drawIndirectCount`.
> * AMD cards going back to the HD 8000 series support `drawIndirectCount`.
> 
> For Android:
> 
> * `drawIndirectCount` is supported on 100% of devices that support Vulkan 1.2.
> * `drawIndirectCount` is supported, as an extension, on 28% of devices that do not support Vulkan 1.2.
> 
> 
> ### Non-zero `firstInstance`
> 
> Requires the `firstInstance` property of the `VkDrawIndirectCommand` and `VkDrawIndexedIndirectCommand` to be non-zero.  This requires the `drawIndirectFirstInstance` feature which is supported on:
> 
> * 99% of desktop GPUs
> * 64% of Android devices
> 
> 
> ## DX12
> 
> All required features are core to DX12.
> 
> 
> ### Multi-Draw
> 
> Uses `ExecuteIndirect` where the `MaxCommandCount` argument is greater than 1 and the `pArgumentBuffer` argument points to a GPU buffer containing an array of `D3D12_DRAW_ARGUMENTS` or `D3D12_DRAW_INDEXED_ARGUMENTS`.
> 
> _NOTE: The binary layout of these structs are compatible with Vulkan._
> 
> 
> ### Draw Count
> 
> Uses `ExecuteIndirect` where the `pCountBuffer` argument is not NULL.  
> 
> 
> ### Non-zero `firstInstance`
> 
> This is the `StartInstanceLocation` of the `D3D12_DRAW_ARGUMENTS` or `D3D12_DRAW_INDEXED_ARGUMENTS` structures.  Has native support for values greater than 0.
> 
> 
> ## Metal
> 
> ### Multi-Draw
> 
> Can be emulated with Indirect Command Buffers (ICBs) and an extra compute shader invocation to translate from the Vulkan-like indirect draw buffer to an ICB.
> 
> Requires
> 
> * iOS 12.0+
> * macOS 10.14+
> * MTLGPUFamilyMac2
> 
> 
> ### Non-zero `firstInstance`
> 
> Natively supported with the `baseInstance` argument.
> 
> 
> ### Draw Count
> 
> Don't record commands past this count in the ICB and use `optimizedIndirectCommandBuffer`.
> 
> Requires
> 
> * iOS 12.0+
> * macOS 10.14+
> * MTLGPUFamilyMac2

**Comment by kainino0x** [comment]
See #4349 for gpuinfo-vulkan-query results and Nov 2023 meeting minutes also.

**Comment by alecazam** [comment]
Just to be clear MDI isn't universal.  It's available on Vulkan desktop and that's it.  On Android, Mali Vulkan is limited to 1 draw from the buffer per draw call.   And much of the older Android vulkan hw.   And on Metal, it's the same.   For some reason, Apple decided to hobble the MDI API on A9 and encourage use of a different more complex API that has range support, but is far more challenging to encode on the gpu.  That encoding support didn't happen until A11.   There's almost no universal support for the indirect count that is also needed.

**Comment by JesseRMeyer** [comment]
There's two concerns at play.  The user API and the underlying implementation.  For a user to specify an array of draw parameters is from an application design perspective already a win towards future proofing if the underlying implementation has to run under emulation, and it's probably no worse in that case than the user doing essentially the same work of issuing a draw per array item in a loop anyway.  The question seems to be related to performance expectations (non-emulated vs accelerated).  But again, for hardware that doesn't natively accelerate MDI would be fed draw calls via a loop of some kind anyway, whether from a user application or the graphics library.  So offering the API generally seems like a net win to me all around.

**Comment by lcodes** [comment]
Behind a feature flag passed at device creation time, this would be perfect.

If it's there, use it, if not fallback to single draws in a loop.

I think it's more valuable to offer advanced APIs as optional features, than not offering them at all.

**Comment by alecazam** [comment]
There is no emulation of a GPU driven count with MDI.  There is the fixed count from the cpu, and the potential for a dynamic count from the GPU.

**Comment by JesseRMeyer** [comment]
The implementation _could_ issue a hard gpu -> cpu sync in that case but I think most (including myself) would prefer an error instead.  Presumably there could be MultiDrawCount and MultiDrawIndirectCount, where the latter is restricted behind a feature flag.

**Comment by mrshannon** [comment]
> There is no emulation of a GPU driven count with MDI. There is the fixed count from the cpu, and the potential for a dynamic count from the GPU.

Actually, you can use a compute shader to emulate this fairly easily by zeroing out the `instanceCount` of the draws after the draw count in the GPU side buffer. If you have hardware backed MDI but not draw count this is not too bad a performance hit. The problem is if you also have to emulate multi-draw itself by running a loop you have to issue many draw calls that may not do anything. So, it's possible to always emulate this approach, the problem is that while it's possible to emulate for correctness, the performance difference between full emulation and no emulation (or draw count emulation) is large. Therefore, it's probably better that this emulation be left to the developer so they can try to do better than naive emulation.

**Comment by logankaser** [comment]
> Therefore, it's probably better that this emulation be left to the developer so they can try to do better than naive emulation.

Worth pointing out that for many years (I'm not familiar with the current state of the codebase),
[multidraw](https://developer.mozilla.org/en-US/docs/Web/API/WEBGL_multi_draw) in chrome did just loop over the 
input. This did actually still help a lot, as it reduced JS vm trampoline and allowed some validation to be hoisted out
of the loop. Just my two cents.

**Comment by alecazam** [comment]
That doesn't work.  This isn't cpu driven.  The MDI data must reside in a buffer, then compute or vertex shader can modify the values, and then the draw needs to increment 1 buffer slot on iOS/Mali/older Android to the cpu count, or by multidraw indirect set by the gpu into a buffer.   The idea of offering two apis with multidraw count indirect optional is feasiable.  iOS and older Android just wouldn't have that.   Also the iOS MDI API can't have additional data (no stride), but Vulkan can.

**Comment by lcodes** [comment]
dispatchCount and indirectCount aren't always easy to emulate, or possible.

There's a hack I'm using by doing a mod/div over the instanceIndex or globalInvocationId, but when the counts differ from batch to batch it's no longer possible.

For example, doing GPU culling means a series of indirect draws each using a different instanceCount and startInstance for each draw.

Doing a naive loop over drawIndirect on the CPU right now, but also issuing 1500 draw calls instead of 1, most of which are zero'd out. It's 1ms of javascript time being wasted.

Having optional support behind a feature flag means users whose hardware support it can handle exponentially larger scenes, while those without support also see reduced simulation limits but can still run it.

I think the need for all target hardware to support a given feature is a downgrade, because it forces high-end machines to the level of the lowest common denominator. More feature flags makes webgpu also more attractive to developers with higher-end hardware in mind.

**Comment by Dampfwalze** [comment]
I want to note that `wgpu` does already implement this draw method. It is behind the [`MULTI_DRAW_INDIRECT`](https://docs.rs/wgpu/latest/wgpu/struct.Features.html#associatedconstant.MULTI_DRAW_INDIRECT) feature, which introduces [`multi_draw_indirect`](https://docs.rs/wgpu/latest/wgpu/struct.RenderPass.html#method.multi_draw_indirect) and [`multi_draw_indexed_indirect`](https://docs.rs/wgpu/latest/wgpu/struct.RenderPass.html#method.multi_draw_indexed_indirect). It is implemented for Vulkan and DX12 and emulated on Metal using `draw_indirect`. But I suspect this support could be extended.

On top of that, it also has the [`MULTI_DRAW_INDIRECT_COUNT`](https://docs.rs/wgpu/latest/wgpu/struct.Features.html#associatedconstant.MULTI_DRAW_INDIRECT_COUNT) feature, which allows you to specify the count via a GPU buffer.

**Comment by lcodes** [comment]
Oh it has push constants too, nice.

Let's get these on the web too :)

**Comment by alecazam** [comment]
With web you get LCD.   Push constants are problematic in other ways.  Some support fp16 and some don't (AMD).  Maybe WebGPU doesn't even offer fp16 support.

**Comment by mrshannon** [comment]
> Push constants are problematic in other ways. Some support fp16 and some don't (AMD). Maybe WebGPU doesn't even offer fp16 support.

I don't need f16 push constants. If I can write u32s push constants, then I can encode/decode anything I need.

**Comment by Kangz** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2024-03-27#drawindirectcount-1354">GPU Web WG 2024-03-27</a></summary>

* KN: accidentally filed another issue about this. Last time we discussed was last Nov, 2023. Milestone prioritization? An optional feature. If we put it in the spec, don't need to implement immediately. But it's complicated.
* KG: Moz would prefer for M2. But we don't have regular indirect support yet - don't want to build on it before we have impl feedback.
* KN: suspect wgpu already has this.
* KG: not the validation.
* JB: can't enable in Firefox because we need to add that validation pass.
* KN: pretty big thing - our team isn't that excited about implementing this either, at least as much as users want it. No problem with M2.

</details>

**Comment by ivanpopelyshev** [comment]
I'm on the way to support webgpu on `tesera.io` . I already use draft `multidraw_bvbi` webgl functionality with fallback on `vertexAttribPointer`, and since this moment I wait for webgpu multidraw too.

God bless whoever implements this!

**Comment by kainino0x** [comment]
Chrome Canary should have `chromium-experimental-multi-draw-indirect` on supported platforms with `chrome://gpu#enable-unsafe-webgpu` enabled.

For anyone who is interested, please try it out and report back! For known Chromium issues, see: https://crbug.com/369246557/dependencies (and please file any bugs you find at https://crbug.com/dawn/new)

If you'd like to help move this feature forward, the most important way to contribute would be by writing conformance tests; please chime in here: https://github.com/gpuweb/cts/issues/3961

**Comment by beaufortfrancois** [comment]
> Chrome Canary should have `chromium-experimental-multi-draw-indirect` on supported platforms with `chrome://gpu#enable-unsafe-webgpu` enabled.

I believe you meant `chrome://flags/#enable-unsafe-webgpu`, not `chrome://gpu#enable-unsafe-webgpu`

**Comment by kainino0x** [comment]
My bad, thank you!

**Comment by RenaudRohlinger** [comment]
Requesting the feature `chromium-experimental-multi-draw-indirect` currently breaks all WebGPU examples in Three.js, throwing the following error regardless of whether the code utilizes indirect command buffers:
```js
Error creating pipeline state Fragment shader cannot be used with indirect command buffers
    at InitializeImpl (..<URL>)
```

Live example failing:
https://rawcdn.githack.com/renaudrohlinger/three.js/utsubo/feat/webgpu-multidrawsupport/examples/webgpu_loader_gltf.html

/cc @kainino0x

**Comment by beaufortfrancois** [comment]
Thanks @RenaudRohlinger for reporting this issue!

I can reproduce in Chrome Canary.  It seems like the "Fragment shader cannot be used with indirect command buffers" error is caused by the [following code](https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/src/dawn/native/metal/RenderPipelineMTL.mm;l=371;drc=b5c7b77c777719f62ebd164a95951a8817cff462) which was added in https://dawn-review.googlesource.com/c/dawn/+/203934:

```cpp
    // Only put this flag on if the feature is enabled because it may have a performance cost.
    descriptorMTL.supportIndirectCommandBuffers =
        GetDevice()->HasFeature(Feature::MultiDrawIndirect);
```

@Sirtsu55 Does that ring a bell by any chance?

**Comment by Sirtsu55** [comment]
Yes it does, That descriptor field wasn't docunented well on the Metal API reference [](https://developer.apple.com/documentation/metal/mtlrenderpipelinedescriptor/2966638-supportindirectcommandbuffers). We assumed that it would be fine to set that property to true since there was nothing explicitly saying it has side effects.

**Comment by Kangz** [comment]
Please let's keep this issue focused on the specification and file browser issues on their respective trackers (in this case https://issues.chromium.org )

**Comment by kainino0x** [comment]
Filed Chromium bug: https://crbug.com/379656378
and tracking thread in case spec work is needed: https://github.com/gpuweb/gpuweb/pull/2315/files#r1847216400

**Comment by chingham** [comment]
Hi ! 5 years later, are there any news on this ?
I know this is available on chrome only, on some devices, under an unsafe flags, but... that's it ?

**Comment by kainino0x** [comment]
There are a number of issues in Dawn (Chromium)'s implementation but the largest one is that I think it is blocked on our migration to Metal Argument Buffers. I just started working on that migration finally so we'll be one step closer.

**Comment by kainino0x** [comment]
@RenaudRohlinger re: https://github.com/gpuweb/gpuweb/pull/2315#issuecomment-4259276305

The Chromium/Dawn tracking issue is https://crbug.com/369246557 and while it isn't currently highly prioritized, thank you for reports of how it will be useful, as it helps us bump it up.

I'm working on the [switch](https://crbug.com/363031535) to argument buffers (which is required as noted in your screenshot) and it's progressing well, but on hold due to the volume of LLM-assisted vulnerability reports.

As an easier alternative it would be possible for us to do something akin to `WEBGL_multi_draw` which is CPU-side, but I'd like to ask if WebGPU render bundles, which we already have, would work as an alternative? They should provide improvements similar to `WEBGL_multi_draw` in places where it's possible to use them.

**Comment by RenaudRohlinger** [comment]
Thanks, I appreciate the update @kainino0x.

To clarify, the goal is not just to reduce the CPU cost of draw submission in a WebGL-style pipeline. The objective is to enable a truly modern, GPU-driven renderer that takes full advantage of WebGPU.

A `WEBGL_multi_draw`-style approach would only help marginally, and similar CPU-side savings can already be approximated with render bundles.

What we actually need is indirect multi-draw with GPU-generated argument buffers, including the ability to source firstInstance from buffer data. That is the primitive needed for a fully GPU-driven pipeline, where culling, compaction, and batching happen in compute, and the render pass consumes the generated indirect arguments directly.

That distinction is critical for workloads like Nanite-style rendering. In my case, the implementation is already working; the remaining blocker is the CPU cost of submitting roughly 100k indirect draws. Render bundles can reduce CPU overhead, but they do not solve the core problem because the draw structure is still defined on the CPU ahead of execution. Without GPU-driven indirect argument consumption, they are not a substitute for true indirect multi-draw.

So from my perspective, the goal should be a proper modern solution that works reliably across browsers through WebGPU, rather than another WebGL-era workaround that improves some cases without actually unlocking fully GPU-driven rendering. I believe this would unlock an entirely new level of Web 3D graphics (Nanite and such).

**Comment by logankaser** [comment]
I work in industry on web map renderers. Map objects are so simple that often CPU culling is a net loss to performance. We can only barely make it win for our more complex objects through an overengineered SIMD-accelerated WASM function operating on sphere bounds.
GPU-driven culling, and draw call overhead reduction would be a game changer for battery life and performance, and it's so much simpler than alternatives. Our native renderer does it, and it makes a huge difference

**Comment by jetp250** [comment]
I suppose I'll leave a comment too, as one of possibly very many who check this thread every so often but don't really have a commercial/non-hobby project to provide substantial weight with: currently, as far as I understand, if I want to write a _modern_ game that runs both natively and on the web (which can be a huge benefit in early testing or demo builds to get visibility since to download & run executables is a much taller ask), right now, I'll be torn primarily between WebGL2's lack of compute shaders and WebGPU's lack of indirect multi-draw capabilities. Both are common in modern graphics for both performance and simplicity, even if you're not writing a bleeding-edge nanite-style renderer like described above. I consider this to be the most critical missing piece from WebGPU, personally, as the single biggest missing piece for portability of modern native renderers to the web.

Also, I'm slightly concerned about that extra compute pass to convert from Vulkan/OpenGL-style draw buffers to Metal ICBs, as on the web there is of course no way to work around this overhead by writing a Metal-specific fast path unlike in native builds, and I personally doubt additional relevant APIs are likely to arise in either Metal or WebGPU later on. Has the option(?) of not exposing the draw command structures directly and instead providing opaque functions to write them, similar to how writing to a Vulkan command buffer works, discussed? For relatively few indirect commands it is a non-issue, but where MDI(C) shines is especially large amounts of draw commands, where that overhead is also higher. How much in practice however, unsure.

Much thanks for everyone's work on this.

**Comment by matthargett** [comment]
I've been bringing WebGPU to lower-end devices including iPhone XS, Apple Watch SE2, and Meta Quest 1 for use in BabylonNative and a custom WebAssembly runtime. In the BabylonNative integration ( https://github.com/rebeckerspecialties/BabylonNative/tree/wgpu ), I did deep profiling on-device to try and get the famous AR Portal demo running on a 60fps camera feed and trying to get the 3D rendering projection to match the framerate. I was only able to get achieve nearly-locked 60fps on iPhone 12 using my implementation of multi-draw indirect on Metal ( https://github.com/gfx-rs/wgpu/pull/9640 ). while iPhone XS didn't reach 60fps, the 1% lows, peak FPS, CPU usage, and battery power draw (almost entirely on were all notably improved. on M4 MacBook Pro, a separate complex rendering + compute workload was able to saturate the display's native refresh rate of 120fps while pinned to CPU efficiency cores, without MDI it hovers around 80fps.

in my opinion, MDI is one of the key features for making web-based 3D rendering actually competitive with native. this was a key concept in maximizing visual splendor and performance on console hardware and game engine projects I worked on professionally.
